### DISCUSSION TITLE
Applying Operations to Filtered Elements in Java Streams with Lambdas

### DISCUSSION THREAD

**Discussion by IsuruSankhajith** [community/community#111396] [Programming Help]
### Body

Is there a concise way to combine filtering and a specific operation within a Java Stream using Lambdas?  For example, if I have a list of products and want to filter for products with a price above $100 and then apply a 10% discount to each matching product's price, what's the most efficient way to achieve this using Streams?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by itshivams** [comment]
### Combining Filtering and Operations in Java Stream with Lambdas

To efficiently combine filtering and a specific operation within a Java Stream using Lambdas, you can leverage the `filter` and `map` functions. For the given scenario of filtering products above $100 and applying a 10% discount, you can use the following approach:

```java
List<Product> discountedProducts = products.stream()
        .filter(product -> product.getPrice() > 100)
        .map(product -> {
            product.setPrice(product.getPrice() * 0.9); // Apply a 10% discount
            return product;
        })
        .collect(Collectors.toList());
```

This code snippet first filters the products based on the price condition and then applies a 10% discount to each matching product. The resulting discountedProducts list contains the modified products.

**Answer:**

Certainly! In Java Streams, you can efficiently combine filtering and specific operations using the filter and map functions. For instance, consider a scenario where you have a list of products and want to filter for those with a price above $100, applying a 10% discount to each matching product's price. Here's a concise way to achieve this:

```java
List<Product> discountedProducts = products.stream()
        .filter(product -> product.getPrice() > 100)
        .map(product -> {
            product.setPrice(product.getPrice() * 0.9); // Apply a 10% discount
            return product;
        })
        .collect(Collectors.toList());
```

In this code snippet:

- The filter operation selects products with a price greater than $100.
- The map operation modifies each matching product by applying a 10% discount to its price.
- The final collect operation gathers the modified products into a list (discountedProducts).

This concise approach allows you to filter and transform the elements of the stream in a single stream pipeline. Adjust the code according to your specific class and attribute names.

**Comment by ngljcb** [comment]
You can filter the products based on the price condition and then apply the discount to the filtered products. Here's how you can do it:

```
import java.util.List;
import java.util.stream.Collectors;

public class Product {
    private String name;
    private double price;

    // Constructor, getters, setters
    // Assume they are implemented
}

public class Main {
    public static void main(String[] args) {
        List<Product> products = /* your list of products */;

        List<Product> discountedProducts = products.stream()
                .filter(product -> product.getPrice() > 100) // Filter products with price above $100
                .map(product -> { // Apply discount of 10% to each product's price
                    double discountedPrice = product.getPrice() * 0.9; // 10% discount
                    product.setPrice(discountedPrice); // Update product's price
                    return product;
                })
                .collect(Collectors.toList());

        // Now 'discountedProducts' contains the filtered products with 10% discount applied
    }
}
```

In this example:

- We use `stream()` to convert the list of products into a stream.
- We then use `filter()` to keep only the products with a price above $100.
- After filtering, we use `map()` to apply a 10% discount to each matching product's price. Inside `map()`, we update each product's price by multiplying it by 0.9.
- Finally, we use `collect(Collectors.toList())` to collect the discounted products into a new list.

This approach combines filtering and applying a specific operation in a concise and efficient manner using Java Streams and lambda expressions.
